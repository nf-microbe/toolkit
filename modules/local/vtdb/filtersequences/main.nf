process VTDB_FILTERSEQUENCES {
    tag "${meta.id}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mulled-v2-80c23cbcd32e2891421c54d1899665046feb07ef:77a31e289d22068839533bf21f8c4248ad274b60-0' :
        'biocontainers/mulled-v2-80c23cbcd32e2891421c54d1899665046feb07ef:77a31e289d22068839533bf21f8c4248ad274b60-0' }"

    input:
    tuple val(meta) , path(fasta)
    tuple val(meta2), path(faa)
    tuple val(meta5), path(completeness_data)
    path(sequences_to_keep)

    output:
    tuple val(meta), path("${prefix}_filtered.fasta.gz")        , emit: fasta
    tuple val(meta), path("${prefix}_filtered_proteins.faa.gz") , emit: proteins
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def seqs_to_keep = sequences_to_keep ? "--seqs_to_keep ${sequences_to_keep}" : '--seqs_to_keep NO_SEQS_TO_KEEP_FILE'
    def faa_args = faa ? "--input_proteins ${faa}" : '--input_proteins NO_PROTEINS_FILE'
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    filtersequences.py \\
        --input_fasta ${fasta} \\
        ${faa_args} \\
        --completeness_data ${completeness_data} \\
        ${seqs_to_keep} \\
        --output_fasta ${prefix}_filtered.fasta \\
        --output_proteins ${prefix}_filtered_proteins.faa \\
        ${args}

    gzip ${prefix}_filtered.fasta ${prefix}_filtered_proteins.faa

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$( python --version | sed 's/Python //' )
        pandas: \$(echo \$(pandas_version.py 2>&1))
        biopython: \$(echo \$(biopython_version.py 2>&1))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}_filtered.fasta.gz
    echo "" | gzip > ${prefix}_filtered_proteins.faa.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$( python --version | sed 's/Python //' )
        pandas: \$(echo \$(pandas_version.py 2>&1))
        biopython: \$(echo \$(biopython_version.py 2>&1))
    END_VERSIONS
    """
}
