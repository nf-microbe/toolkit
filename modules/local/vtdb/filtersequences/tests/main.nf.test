nextflow_process {

    name "Test Process: FILTERSEQUENCES"
    script "../main.nf"
    process "FILTERSEQUENCES"

    test("fasta.gz + faa.gz + combined_data.tsv + sequences_to_keep.tsv") {
        when {
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test.fasta.gz", checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test.faa.gz", checkIfExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test_completeness_data.tsv", checkIfExists: true)
                    ]
                )
                input[3] = Channel.of(
                    [
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test_sequences_to_keep.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("fasta.gz + faa.gz + combined_data.tsv + no sequences to keep") {
        when {
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test.fasta.gz", checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test.faa.gz", checkIfExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test_completeness_data.tsv", checkIfExists: true)
                    ]
                )
                input[3] = Channel.of([])
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("stub") {
        when {
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test.fasta.gz", checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test.faa.gz", checkIfExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id:'test' ],
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test_completeness_data.tsv", checkIfExists: true)
                    ]
                )
                input[3] = Channel.of(
                    [
                        file("${projectDir}/assets/test-datasets/modules/local/filtersequences/test_sequences_to_keep.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
