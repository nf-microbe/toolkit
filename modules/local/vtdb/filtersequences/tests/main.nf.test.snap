{
    "fasta.gz + faa.gz + combined_data.tsv + no sequences to keep": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.fasta.gz:md5,8bb9e936bc24116c8b538ba95af9c482"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered_proteins.faa.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,d4721da6102734d4c962030ed8c0fc17"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.fasta.gz:md5,8bb9e936bc24116c8b538ba95af9c482"
                    ]
                ],
                "proteins": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered_proteins.faa.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d4721da6102734d4c962030ed8c0fc17"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-15T07:27:08.506438928"
    },
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.fasta.gz:md5,8bb9e936bc24116c8b538ba95af9c482"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered_proteins.faa.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,d4721da6102734d4c962030ed8c0fc17"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.fasta.gz:md5,8bb9e936bc24116c8b538ba95af9c482"
                    ]
                ],
                "proteins": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered_proteins.faa.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d4721da6102734d4c962030ed8c0fc17"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-15T07:27:10.554334423"
    },
    "fasta.gz + faa.gz + combined_data.tsv + sequences_to_keep.tsv": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.fasta.gz:md5,8bb9e936bc24116c8b538ba95af9c482"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered_proteins.faa.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,d4721da6102734d4c962030ed8c0fc17"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.fasta.gz:md5,8bb9e936bc24116c8b538ba95af9c482"
                    ]
                ],
                "proteins": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered_proteins.faa.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d4721da6102734d4c962030ed8c0fc17"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-14T16:43:44.702919982"
    }
}